src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/RecommenderPackageFeed.cs

```diff
@@ -0,0 +1,192 @@
+
+            if (logger == null)
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+            _logger = logger;
+
+            // Get NuGet package recommender service, or null if it is not available
+            NuGetRecommender = Package.GetGlobalService(typeof(SVsNuGetRecommenderService)) as IVsNuGetPackageRecommender;
+            if(NuGetRecommender != null)
+            {
+                var VersionDict = NuGetRecommender.GetVersionInfo();
+                VersionInfo.modelVersion = VersionDict.ContainsKey("Model") ? VersionDict["Model"] : null;
+                VersionInfo.vsixVersion = VersionDict.ContainsKey("Vsix") ? VersionDict["Vsix"] : null;
+            }
+        }
+
+        private class RecommendSearchToken : ContinuationToken
+        {
+            public string SearchString { get; set; }
+            public SearchFilter SearchFilter { get; set; }
+        }
+
+        public Task<SearchResult<IPackageSearchMetadata>> SearchAsync(string searchText, SearchFilter searchFilter, CancellationToken cancellationToken)
+        {
+            var searchToken = new RecommendSearchToken
+            {
+                SearchString = searchText,
+                SearchFilter = searchFilter,
+            };
+
+            return RecommendPackagesAsync(searchToken, cancellationToken);
+        }
+
+        public async Task<SearchResult<IPackageSearchMetadata>> RecommendPackagesAsync(ContinuationToken continuationToken, CancellationToken cancellationToken)
+        {
+            var searchToken = continuationToken as RecommendSearchToken;
+            if (searchToken == null)
+            {
+                throw new InvalidOperationException("Invalid token");
+            }
+            // don't make recommendations if the user entered a search string
+            if (searchToken.SearchString != "")
+            {
+                return SearchResult.Empty<IPackageSearchMetadata>();
+            }
+
+            List<string> recommendIds = new List<string>();
+            if (NuGetRecommender != null)
+            {
+                // get lists of only the package ids to send to the recommender
+                List<string> topPackages = _installedPackages.Select(item => item.Id.ToLowerInvariant()).ToList();
+                List<string> depPackages = _dependentPackages.Keys.ToList();
+                // call the recommender to get package recommendations
+                recommendIds = await NuGetRecommender.GetRecommendedPackageIdsAsync(_targetFrameworks, topPackages, depPackages, cancellationToken);
+            }
+
+            // get PackageIdentity info for the top 5 recommended packages
+            int _maxRecommend = 5;
+            int index = 0;
+            List<PackageIdentity> recommendPackages = new List<PackageIdentity>();
+            while (recommendIds != null && index < recommendIds.Count() && recommendPackages.Count < _maxRecommend)
```

Comment:
If `recommendIds` is null, a whole lot of the following logic is irrelevant. Please add an if-statement to wrap what's relevant below.

Reply:
Rearranged the code so that this is only used when there are recommendations.